Generate and test complex infras

eckter opened the pull request:
Fixes #255

eckter committed:
infra-generator: prevent short paths from being inverted

infra-generator: make switch position map use endpoint in the key

eckter committed:
debug-viewer: allow multiple edges between nodes

eckter committed:
infra-generator: make random infras acyclic

infra-generator: generate random train paths

eckter committed:
infra-generator: fix bug where a train could "turn around" in path

eckter committed:
infra-generator: add random speed sections

eckter committed:
tests: add simple parametrized allowance tests on all infras

tests: add generated infras to the margin tests

tests: add docstring to public methods

tests: generated infras: adapt to new test generation

flomonster commented on examples/generator/railjson_generator/schema/infra/endpoint.py:
```diff
@@ -23,6 +23,12 @@ def get_neighbors(self) -> "TrackEndpoint":
+    def opposite(self) -> "TrackEndpoint":
+        if self.endpoint is Endpoint.BEGIN:
+            return TrackEndpoint(self.track_section, Endpoint.END)
+        else:
+            return TrackEndpoint(self.track_section, Endpoint.BEGIN)
```
```suggestion
        return TrackEndpoint(self.track_section, Endpoint.BEGIN)
```

flomonster commented on examples/generator/railjson_generator/schema/infra/infra.py:
```diff
@@ -20,6 +20,7 @@ class Infra:
     operational_points: List[OperationalPoint] = field(default_factory=list)
     tvd_sections: List[TVDSection] = field(default_factory=list)
     routes: List[Route] = field(default_factory=list)
+    speed_limits: List[Dict] = field(default_factory=list)
```
Do we need to specify the label of a speed limit, we could generate it from speed limits in track sections.

flomonster commented on examples/generator/railjson_generator/schema/infra/track_section.py:
```diff
@@ -29,6 +29,9 @@ class TrackSection:
     end_links: List[TrackEndpoint] = field(default_factory=list, repr=False)
+    slopes: List[Tuple[float, float, float]] = field(default_factory=list)
+    curves: List[Tuple[float, float, float]] = field(default_factory=list)
+    speed_limits: List[Tuple[float, float, str]] = field(default_factory=list)
```
I would have added multiple datastruct such as `RangeElement`, `Slope`, `Curve` and `SpeedLimit`  to stay consistent with the rest of the library.

flomonster commented on examples/generator/railjson_generator/schema/infra/track_section.py:
```diff
@@ -91,8 +103,21 @@ def format(self):
-            "curves": [],
-            "speed_sections": [],
             "endpoints_coords": endpoints_coords,
+            "slopes": [{
+                "begin": begin,
+                "end": end,
+                "gradient": gradient,
+            } for begin, end, gradient in self.slopes],
+            "curves": [{
+                "begin": begin,
+                "end": end,
+                "radius": radius,
+            } for begin, end, radius in self.curves],
+            "speed_sections": [{
+                "begin": begin,
+                "end": end,
+                "ref": section_id,
+                "applicable_direction": "BOTH"
```
All of this could be simplify by calling a `format()` method

eckter committed:
infra generator: change tuples to classes

infra-generator: fix routes going before entry point

eckter committed:
tests: change 3 trains TST test to use any switch id

tests: ignore signalization in margin tests

infra-generator: avoid loops in schedules
